### PR TITLE
search: scaffolding for combinatorial query generator

### PR DESCRIPTION
**Pull request by rvantonder**
Hokay started hacking on this because I'm trying to rush to the end MVP state.

This adds a generator that only takes a flat set of individual rules and tries combinations of `n choose k` on the rule set, for decreasing values of `k`. See the test and docstring for how this works.

This is unused, but it will replace the current `Generator` code which only tries to apply the list of rules, not combinations of individual ones. When the code uses the new generator, it should turn out that we can just add individual rules and leave the iteration to try combinations that apply on the query. This'll make it much easier for me to add rules that, e.g., interpret `https://github.com/repo/repo` as a repo filter without having to meddle with composition of rules. This isn't the only strategy we have to use, it's just one that will work well for me right now.

This still needs some other tweaking to incorporate the unordered patterns rule. More on that to come later.

## Test plan
Unused, adds test.

### CODE REVIEW

File: internal/search/job/jobutil/feeling_lucky_search_job.go [hunk 1]
```diff
@@ -565,3 +566,92 @@ func (g *generatedSearchJob) MapChildren(fn job.MapFunc) job.Job {
...
+type cg = *combin.CombinationGenerator
+
+// NewComboGenerator returns a generator for queries produces by a combination
+// of rules on a seed query. The generator strategy tries to first apply _all_
+// rules, and then successively reduces the number of rules that it attempts to
+// apply by one. This strategy is useful when we try the most aggressive
+// interpretation of a query subject to rules first, and gradually loosen the
+// number of rules and interpretation. To avoid spending time on generator
+// invalid combinations, the generator prunes the initial rule set to only those
+// rules that do successively apply individually to the seed query.
```
Comment (rvantonder) [inline internal/search/job/jobutil/feeling_lucky_search_job.go:599]: It would be better to "prune as we go", but that's significantly more complex, so I will just prune beforehand for `number of rules ~= 10`. once it gets larger than that it'll be time to try inline the pruning.

File: internal/search/job/jobutil/feeling_lucky_search_job_test.go [hunk 2]
```diff
@@ -169,3 +169,38 @@ func TestGeneratedSearchJob(t *testing.T) {
...
+
+		if g == nil {
+			break
+		}
+	}
+
+	result, _ := json.MarshalIndent(generated, "", "  ")
+
+	t.Run("ok", func(t *testing.T) {
+		autogold.Equal(t, autogold.Raw(result))
```
Comment (tbliu98) [inline internal/search/job/jobutil/feeling_lucky_search_job_test.go:204]: Will rules/combinations of rules always be applied in the same order?

**Review by tbliu98 — APPROVED** [review]

File: internal/search/job/jobutil/feeling_lucky_search_job.go [hunk 3]
```diff
@@ -565,3 +566,91 @@ func (g *generatedSearchJob) MapChildren(fn job.MapFunc) job.Job {
...
+		description: "apply search type for pattern",
+		transform:   typePatterns,
+	},
+	{
+		description: "apply language filter for pattern",
+		transform:   langPatterns,
+	},
+}
+
+type cg = *combin.CombinationGenerator
```
Comment (camdencheek) [inline internal/search/job/jobutil/feeling_lucky_search_job.go:590]:
```suggestion
type cg *combin.CombinationGenerator
```

File: internal/search/job/jobutil/feeling_lucky_search_job.go [hunk 4]
```diff
@@ -565,3 +566,91 @@ func (g *generatedSearchJob) MapChildren(fn job.MapFunc) job.Job {
...
+				}
+
+				q := autoQuery{
+					description: strings.Join(descriptions, " and "),
+					query:       *generated,
+				}
+
+				return &q, n(k, cg)
+			}
+			k -= 1
```
Comment (camdencheek) [inline internal/search/job/jobutil/feeling_lucky_search_job.go:633]: So, if I'm understanding this correctly, we start by applying _all_ the rules, then we reduce the number of rules applied by one each time. This would probably work well when we have a reasonable expectation that the rules generate a better query than what the user gave us (e.g. unquote), but I'm not sure it makes sense when applying more rules increases deviation from the user's original query. 

I get the impression that we'll get higher quality results if we start by applying one rule, then increase the number of rules applied as we go so the results returned first are those that come from queries that have the smallest deviation from the original query.

**Review by camdencheek — COMMENTED** [review]
Sorry for the post-review. A couple of small comments, nothing critical.

File: internal/search/job/jobutil/feeling_lucky_search_job.go (hunk 3 above)
Comment (rvantonder) [reply internal/search/job/jobutil/feeling_lucky_search_job.go:590]: nope, actually really do want this. it doesn't type check otherwise 

> ./feeling_lucky_search_job.go:606:9: cg.Next undefined (type cg has no field or method Next)                                                                                                                                           
./feeling_lucky_search_job.go:609:27: cg.Combination undefined (type cg has no field or method Combination)  

(I just want an alias so I don't have to put this long thing in the signature)

Comment (camdencheek) [reply internal/search/job/jobutil/feeling_lucky_search_job.go:590]: Aaah, gotcha. That makes sense. Was thinking you were just reusing the data structure.

File: internal/search/job/jobutil/feeling_lucky_search_job.go (hunk 4 above)
Comment (rvantonder) [reply internal/search/job/jobutil/feeling_lucky_search_job.go:633]: very much want and appreciate your feedback. Yes, so now we are getting into interesting territory which like, it's awesome if other people have ideas and want to explore. I'm making this choice based off of what I'm seeing with the current rules.

The main reason I want to try all rules first is because it allows rules to introduce "shorthand" for queries. For example, if you map `commit` to `type:commit` and `go` to `lang:go`, then you can type:

`commit go blah` and you will get automatic expansion of `commit` and `go` first. I've personally been relying on the shorthand expansion a lot and find it really nice, so I want all the `pattern -> filter` expansions to hit right off the bat.

For sure not _all_ rules have this property (I think, especially ones that may not convert `pattern -> filter`). This is something to tune in a generator (you can write your own generator or strategies!) and we can learn what are cool ways to do more.

Comment (rvantonder) [reply internal/search/job/jobutil/feeling_lucky_search_job.go:633]: so to your point, applying all of _these kinds_ of rules does deviate further from the original query, but they are more specific with respect to certain pattern keywords. Naturally not all of _these kinds_ of rules will apply if they don't fire (e.g., you don't have a `commit` keyword in the query, or a value like `go` that maps to `lang`).

Comment (rvantonder) [reply internal/search/job/jobutil/feeling_lucky_search_job.go:633]: yeah, so another way to look at this, for these kinds of rules, when they apply, is in terms of the anticipated result set _size_. Apply more of these rules narrow the specificity of the query, and the intuition is that that result set is _smaller_. E.g.,:

`go pattern some/directory`

The result set is going to be smaller if we do:

- `lang:go path:some/directory pattern`

than

- `lang:go some/directory pattern`

and the intuition is that I would want the smaller (more specific) result set first. So for these kinds of rules: greater initial deviation, but that implies smaller result sets.

You might like the next PR, which introduces rules that (by my intuition) _increases_ the result set size. So: `parse func` -> `parse AND func` is an example where I expect a _larger_ result set size. We don't want to try that _first_ because that result set size might dwarf the more specific, smaller result sets.

Basically there is clearly a rationale for having different kinds of rules, and right now I'm constructing it so that you have narrowing rules (apply as many as possible) and widening rules (gradually apply on top of narrow rules after trying all subsets of narrowing rules)
